add-custom-dynamic-entity-data-source/src/main/java/com/esri/arcgismaps/sample/addcustomdynamicentitydatasource/components/MapViewModel.kt

```diff
@@ -148,28 +162,29 @@ class MapViewModel(application: Application) : AndroidViewModel(application) {
                 maximumResults = 1
             ).onSuccess { result ->
                 (result.geoElements.firstOrNull() as? DynamicEntityObservation)?.let { observation ->
-                    // Get the dynamic entity from the observation.
-                    observation.dynamicEntity?.let { dynamicEntity ->
-                        // Build a string for observation attributes.
-                        val stringBuilder = StringBuilder()
-                        for (attribute in arrayOf("VesselName", "CallSign", "COG", "SOG")) {
-                            val value = dynamicEntity.attributes[attribute]?.toString()
-                            // Account for when an attribute has an empty value.
-                            if (!value.isNullOrEmpty()) {
-                                stringBuilder.appendLine("$attribute: $value")
-                            }
+                    // Set the identified dynamic entity, used to display the callout.
+                    selectedGeoElement = observation.dynamicEntity
+                    // Define a new CoroutineScope to collect observation events on.
+                    observationsJob = launch(Dispatchers.IO) {
+                        // Collect observation events and update the observation string accordingly.
+                        observation.dynamicEntity?.dynamicEntityChangedEvent?.collect { dynamicEntityChangedInfo ->
+                            // Parse the observation attributes, filter out empty values, and remove
+                            // starting and ending {}s.
+                            observationString =
+                                dynamicEntityChangedInfo.receivedObservation?.attributes?.filter {
+                                    it.value.toString().isNotEmpty()
+                                }.toString().replaceFirst("{", " ").removeSuffix("}")
+                                    .replace(",", "\n")
                         }
-                        val entityAttributes = stringBuilder.toString().trimEnd()
-                        messageDialogVM.showMessageDialog(
-                            title = "Identified dynamic entity",
-                            description = entityAttributes
-                        )
                     }
+                    // If no observation is found, set the selectedGeoElement to null.
+                } ?: run {
+                    selectedGeoElement = null
                 }
             }.onFailure { error ->
-                messageDialogVM.showMessageDialog(
-                    title = "Error in evaluating popup expression: ${error.message.toString()}",
-                    description = error.cause.toString()
+                Log.e(
+                    javaClass.simpleName,
+                    "Error identifying dynamic entity: ${error.message}"
```

Comment:
For MapViewModel error handling I would recommend using the `MessageDialogViewModel` from samples-lib which will display the error message on the screen and also Log the error.

Reply:
I think this was previously included in the sample but was removed during bug testing the Callout.